plugins/paper/src/main/java/com/playmonumenta/plugins/network/ClientModHandler.java

```diff
@@ -0,0 +1,170 @@
+		if (!playerHasClientMod(player) || !shouldHandleAbility(ability))
+			return;
+		sendPacket(player, prepareAbilityUpdatePacket(player, ability));
+	}
+
+	private AbilityUpdatePacket prepareAbilityUpdatePacket(Player player, Ability ability) {
+		ClassAbility classAbility = ability.getInfo().mLinkedSpell;
+		int remainingCooldown = classAbility == null ? 0 : mPlugin.mTimers.getCooldown(player.getUniqueId(), classAbility);
+		int charges = ability instanceof AbilityWithChargesOrStacks ? ((AbilityWithChargesOrStacks) ability).getCharges() : 0;
+		int maxCharges = ability instanceof AbilityWithChargesOrStacks ? ((AbilityWithChargesOrStacks) ability).getMaxCharges() : 0;
+
+		AbilityUpdatePacket packet = new AbilityUpdatePacket();
+		packet.name = getAbilityName(ability);
+		packet.remainingCooldown = remainingCooldown;
+		packet.initialCooldown = ability.getInfo().mCooldown;
+		packet.remainingCharges = charges;
+		packet.maxCharges = maxCharges;
+		return packet;
+	}
+
+	/**
+	 * Sends a class update to the player.
+	 * <p>
+	 * Does nothing if the player does not have a compatible client mod installed.
+	 */
+	public void updateAbilities(Player player) {
+		if (!playerHasClientMod(player))
+			return;
+
+		AbilityUpdatePacket[] abilities = mPlugin.mAbilityManager.getPlayerAbilities(player).getAbilities().stream()
+			.filter(ClientModHandler::shouldHandleAbility)
+			.map(ability -> prepareAbilityUpdatePacket(player, ability))
+			.sorted(Comparator.comparing(p -> p.name))
+			.toArray(AbilityUpdatePacket[]::new);
+
+		ClassUpdatePacket packet = new ClassUpdatePacket();
+		packet.abilities = abilities;
+		sendPacket(player, packet);
+	}
+
+	public void silenced(Player player, int duration) {
+		PlayerStatusPacket packet = new PlayerStatusPacket();
+		packet.silenceDuration = duration;
+		sendPacket(player, packet);
+	}
+
+	private void sendPacket(Player player, Packet packet) {
+		player.sendPluginMessage(mPlugin, CHANNEL_ID, mGson.toJson(packet).getBytes(StandardCharsets.UTF_8));
+	}
+
+	/**
+	 * @return Whether we're sending data for the given ability to clients
+	 */
+	private static boolean shouldHandleAbility(Ability ability) {
+		return ability != null && (ability.getInfo().mCooldown > 0 || ability instanceof AbilityWithChargesOrStacks);
+	}
+
+	private static String getAbilityName(Ability ability) {
+		// The ClassAbility name is preferable if it exists (e.g. for the two Elemental Spirits)
+		if (ability.getInfo().mLinkedSpell != null) {
+			return ability.getInfo().mLinkedSpell.getName();
+		}
+		return ability.getDisplayName();
+	}
+
+	private static boolean playerHasClientMod(Player player) {
+		return player.getListeningPluginChannels().contains(CHANNEL_ID);
+	}
+
+	// ------ packets ------
+
+	private interface Packet {
+	}
+
+	/**
+	 * Sent whenever a player's class is updated.
+	 */
+	private static class ClassUpdatePacket implements Packet {
+
+		final String _type = "ClassUpdatePacket";
```

Comment:
This field is used as type tag to be able to identify which packet is sent (since they all are plain JSON objects).
I could change Gson to include static fields, but that just feels ugly.

Reply:
Ahh, ok. Didn't realize GSON was picky about statics. Agree, that makes sense.